### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by marksteward** [emfcamp/Website#879]
We hit 100% CPU around 18:59:30 and latency shot up to 7s by 19:00:00, rising to 20s by 19:00:30. We got over 250 500 errors, but these don't show up on prometheus, which is because they were generated in nginx.

Looks like some people were full-refreshing, and even loading the tickets page causes a decent amount of CPU load.

A bunch of people lost their tickets, and we also show a really unhelpful message when the number of available tickets is less than requested.

We need to:

- [x] Add 2 more CPUs
- [x] Reduce CPU usage normally
- [ ] Consider adding queueing
- [ ] Consider requiring logins before ticket sales
- [x] Deal with 500s better - we should link back to the "recover basket" page
- [x] Add ratelimiting
- [x] Fix the "not enough tickets" error, and make sure F5 from a 500 won't do the same
- [x] Check the prometheus endpoint won't collapse under pressure
- [x] Aim locust at the site - we need to cope with 100+ users/s again
- [x] Surface nginx 500s somehow

**Comment by russss** [comment]
Rate limiting is pretty simple with nginx, I'll sort that out https://www.nginx.com/blog/rate-limiting-nginx/

I think queueing is going to require quite a lot of thought.

**Comment by marksteward** [comment]
Very fast full-page reloading:

```
19/Feb/2020:18:57:07 +0000      GET /tickets HTTP/2.0   200     3932    0.041           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:07 +0000      GET /tickets HTTP/2.0   200     3936    0.088           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:07 +0000      GET /tickets HTTP/2.0   200     3930    0.082           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:07 +0000      GET /tickets HTTP/2.0   200     3934    0.088           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:07 +0000      GET /tickets HTTP/2.0   499     0       0.072           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:08 +0000      GET /tickets HTTP/2.0   200     3934    0.075           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:08 +0000      GET /tickets HTTP/2.0   200     3932    0.060           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:08 +0000      GET /tickets HTTP/2.0   200     3934    0.066           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:08 +0000      GET /tickets HTTP/2.0   200     3932    0.046           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:08 +0000      GET /tickets HTTP/2.0   200     3934    0.046           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:08 +0000      GET /tickets HTTP/2.0   200     3936    0.050           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:09 +0000      GET /tickets HTTP/2.0   200     3934    0.051           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:09 +0000      GET /tickets HTTP/2.0   200     3935    0.072           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:09 +0000      GET /tickets HTTP/2.0   200     3936    0.050           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:09 +0000      GET /tickets HTTP/2.0   200     3934    0.050           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:10 +0000      GET /tickets HTTP/2.0   200     3933    0.048           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:11 +0000      GET /tickets HTTP/2.0   200     3939    0.047           https://www.emfcamp.org/tickets Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.
19/Feb/2020:18:57:20 +0000      GET /tickets HTTP/2.0   200     3932    0.076           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:21 +0000      GET /tickets HTTP/2.0   200     3930    0.143           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:22 +0000      GET /tickets HTTP/2.0   200     3935    0.088           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:22 +0000      GET /tickets HTTP/2.0   200     3935    0.065           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:22 +0000      GET /tickets HTTP/2.0   200     3933    0.045           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:23 +0000      GET /tickets HTTP/2.0   200     3934    0.047           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:23 +0000      GET /tickets HTTP/2.0   200     3933    0.150           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:23 +0000      GET /tickets HTTP/2.0   200     3934    0.063           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:23 +0000      GET /tickets HTTP/2.0   200     3933    0.070           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:29 +0000      GET /tickets HTTP/2.0   200     3933    0.057           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:30 +0000      GET /tickets HTTP/2.0   200     3934    0.045           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:30 +0000      GET /tickets HTTP/2.0   200     3934    0.113           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:30 +0000      GET /tickets HTTP/2.0   499     0       0.159           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:30 +0000      GET /tickets HTTP/2.0   200     3933    0.130           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
19/Feb/2020:18:57:30 +0000      GET /tickets HTTP/2.0   200     3935    0.053           -       Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, l
```

These all included static assets.

**Comment by marksteward** [comment]
<img width="1246" alt="Screenshot 2020-02-19 at 19 58 03" src="https://user-images.githubusercontent.com/121960/74877024-a7785800-535c-11ea-8b7c-8a4c3b3d4e29.png">
<img width="1244" alt="Screenshot 2020-02-19 at 19 58 05" src="https://user-images.githubusercontent.com/121960/74877031-a9dab200-535c-11ea-9206-c8559b55d31b.png">
<img width="1243" alt="Screenshot 2020-02-19 at 19 58 08" src="https://user-images.githubusercontent.com/121960/74877034-aa734880-535c-11ea-8fa2-d9ed157c7f72.png">

**Comment by russss** [comment]
I've added request rate, request time, and error rate graphs to Grafana from the nginx logs. Now we have these, I'm not especially fussed about the Prometheus endpoint going down briefly, I think we can ignore that.

**Comment by russss** [comment]
I fixed a significant SQL performance issue in the tickets page with 82d7ef5.

**Comment by russss** [comment]
After a not inconsiderable amount of faff, we're also now regression-testing the number of queries issued by the tickets page in [tests/test_sql_query_count.py](https://github.com/emfcamp/Website/blob/master/tests/test_sql_query_count.py).

**Comment by russss** [comment]
The "not enough tickets" error is fixed, I believe, by 3bbfcbd. If we can't assign enough tickets we will reserve as many as we can and send the user to the payment-choose page.

I've also made some wording improvements to the payment-choose page which will hopefully reassure people that their tickets are reserved and reduce the chances of people creating payments in the wrong currency.

**Comment by marksteward** [comment]
Much better!

<img width="1235" alt="Screenshot 2020-03-14 at 11 13 45" src="https://user-images.githubusercontent.com/121960/76680945-42272800-65e5-11ea-801f-9a60a5d1cb0a.png">
<img width="1236" alt="Screenshot 2020-03-14 at 11 13 49" src="https://user-images.githubusercontent.com/121960/76680946-45221880-65e5-11ea-8ac8-3efb717fafd4.png">
<img width="1237" alt="Screenshot 2020-03-14 at 11 13 52" src="https://user-images.githubusercontent.com/121960/76680947-45baaf00-65e5-11ea-8326-58f69b372b6b.png">
